opt: Integrate initial versions of v4 expr, memo, and factory

andy-kimball opened the pull request:
This PR takes a subset of the files from the v4 opttoy branch, adds
comments and tests, and puts them in a new opt/xform package.
Ultimately, I'd expect us to have packages like:

  opt/build - contains AST => v4 Expr builder
  opt/xform - transformation and optimization of v4 Exprs
  opt/exec  - contains v4 Expr => plan nodes (is exec best name?)

The testing of this PR is not complete. I'll be adding more tests
once the builder is in place, so that I can use SQL. For now, this
PR does some rudimentary testing using the memo factory directly.

andy-kimball committed:
opt: Integrate initial versions of v4 expr, memo, and factory

This PR takes a subset of the files from the v4 opttoy branch, adds
comments and tests, and puts them in a new opt/xform package.
Ultimately, I'd expect us to have packages like:

  opt/build - contains AST => v4 Expr builder
  opt/xform - transformation and optimization of v4 Exprs
  opt/exec  - contains v4 Expr => plan nodes (is exec best name?)

The testing of this PR is not complete. I'll be adding more tests
once the builder is in place, so that I can use SQL. For now, this
PR does some rudimentary testing using the memo factory directly.

Also, in order to pass the lint test, *.og.go files were excluded
from crlfmt, since they are auto-generated by the new optgen tool.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21478)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=21478) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=21478) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=21478) it.</sub>

RaduBerinde commented:
The new comments help a lot, thanks!  Just some nits from an initial look, I will need to go through it multiple times.

---

Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/expr.go, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-EISNnromcnlk-4Dk:-L3-EISNnromcnlk-4Dl:b-hiu8bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L86)):*
> ```Go
> // makeExpr creates a new Expr instance that references the expression in the
> // specified group that satisfies the required properties at the lowest cost.
> func makeExpr(mem *memo, group GroupID, required physicalPropsID) Expr {
> ```

[nit] The name is not very descriptive, maybe `extractLowestCostExpr`

---

*[pkg/sql/opt/xform/logical_props.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-GSb8haj05RR1hluL:-L3-GSb9n8ulM8JrR3Yk:b-oihv5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props.go#L65)):*
> ```Go
> 			fmt.Fprintf(&buf, "%d", colIndex)
> 			if p.Relational.NotNullCols.Contains(int(colIndex)) {
> 				buf.WriteByte('*')
> ```

[nit] in opt we switched to use `,null` for nullable columns instead (`*` is cryptic)

---

*[pkg/sql/opt/xform/memo_expr.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-Gp4D86is8Idjge5_:-L3-Gp4D86is8Idjge5a:b-q24yea) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L28)):*
> ```Go
> // memoExpr is a memoized representation of an expression. Strongly-typed
> // specializations of memoExpr are generated by optgen for each operator (see
> // Expr.og.go). Each memoExpr belongs to a memo group, which contain logically
> ```

[nit] expr.og.go, contains

---

*[pkg/sql/opt/xform/memo_expr.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HDY3ZXPm5a4iOqaP:-L3-HDY3ZXPm5a4iOqaQ:b-8z988z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L52)):*
> ```Go
> // expressions share the same fingerprint, then they are the identical
> // expression. If they don't share a fingerprint, then they still may be
> // logically equivalent expressions. Since a memo expression is 16 bytes and
> ```

24 bytes (or just "small")

---

*[pkg/sql/opt/xform/memo_expr.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HVO5xU-y8htbglQ3:-L3-HVO5xU-y8htbglQ4:b81n0rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L61)):*
> ```Go
> }
> 
> func (me memoExpr) memoString(mem *memo, mgrp *memoGroup) string {
> ```

[nit] memoGroupString (maybe it should be a method on memoGroup?)

---

*[pkg/sql/opt/xform/physical_props.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-IzGNSp5Ktvo63mxa:-L3-IzGNSp5Ktvo63mxb:b3qiphq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/physical_props.go#L42)):*
> ```Go
> // properties while still remaining logically equivalent.
> type PhysicalProps struct {
> }
> ```

[nit] add a // TODO here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L3-L-9W_TSYUWwKORJJ:b-v5rw1n)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Looks great so far!  I've reviewed some of the files.  I'll come back in a bit to do the others.

---

Reviewed 4 of 15 files at r1.
Review status: 4 of 15 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/expr.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-El3lbVepms7XyB3J:-L3-El3lbVepms7XyB3K:b-n18r79) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L32)):*
> ```Go
> // and examine its properties. Given a particular root memo group and a set of
> // required physical properties, there is exactly one lowest cost tree in the
> // memo forest that provides those properties. Expr overlays the memo groups
> ```

It seems like there could be scenarios where there is a tie (e.g., two join orderings for tables that are exactly the same size).  How are ties broken?

---

*[pkg/sql/opt/xform/expr.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HTU_2yQhvgOscK09:-L3-HTU_2yQhvgOscK0A:bmln1q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L114)):*
> ```Go
> // as the ordering of result rows.
> func (e *Expr) Physical() *PhysicalProps {
> 	return &PhysicalProps{}
> ```

It would help to add a comment to explain that this functionality isn't implemented yet.

---

*[pkg/sql/opt/xform/factory.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-N-4HNTAVVnT9-MOE:-L3-N-4ItfShvii2XYCg:ba87nhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/factory.go#L32)):*
> ```Go
> type Factory struct {
> 	mem      *memo
> 	maxSteps int
> ```

Add a comment to explain what `maxSteps` is.

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-OxjyMkuiFSMXTtl4:-L3-OxjyMkuiFSMXTtl5:b-jeboyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_factory.go#L38)):*
> ```Go
> // NOTE: The parent expression is passed as an Expr for convenient access to
> // 		 children, but certain properties on it are not yet defined (like its
> // 		 logical properties!).
> ```

[nit] fix spacing issues in the NOTE

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 187 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-S6kBBgInFZNbdMyl:-L3-S6kBBgInFZNbdMym:b-7crqig) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_factory.go#L187)):*
> ```Go
> 
> 	// Columns have to be not-null on both sides to be not-null in result.
> 	for leftIndex, rightIndex := range colMap {
> ```

why not just do a simple intersection? e.g. 
```
props.Relational.NotNullCols = leftProps.Relational.NotNullCols.Intersection(rightProps.Relational.NotNullCols)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L3-ToKTkLyU8VvwYA3H:bmwtw57)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 15 files at r1.
Review status: 5 of 15 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 187 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-S6kBBgInFZNbdMyl:-L3-bVpYiZdp2u1PnbyI:bi5fx6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_factory.go#L187)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

why not just do a simple intersection? e.g. 
```
props.Relational.NotNullCols = leftProps.Relational.NotNullCols.Intersection(rightProps.Relational.NotNullCols)
```
</blockquote></details>

Never mind - I was confused about what the colMap was doing. Might be good to add a comment with a reminder/explanation about what colMap is.

---

*[pkg/sql/opt/xform/logical_props_test.go, line 79 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-Ziv4665Mng1UZlYq:-L3-Ziv4665Mng1UZlYr:ba3rw0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_test.go#L79)):*
> ```Go
> 	joinFunc := func(op Operator, expected string) {
> 		// (Join (Scan a) (Scan b) (True))
> 		factory := newFactory(gen.mem, 0)
> ```

when you pass in a constant like 0 or false, it's good practice to add a comment like this:
```
factory := newFactory(gen.mem, 0 /* maxSteps */)
```
It's in the style guide somewhere...

---

*[pkg/sql/opt/xform/logical_props_test.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-_Ahh9-e9uH-5sjB5:-L3-_Ahh9-e9uH-5sjB6:blna65f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_test.go#L83)):*
> ```Go
> 		rightGroup := factory.ConstructScan(factory.InternPrivate(b))
> 		onGroup := factory.ConstructTrue()
> 		joinGroup := factory.DynamicConstruct(op, []GroupID{leftGroup, rightGroup, onGroup}, 0)
> ```

ditto here - add a comment for the 0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 15 files at r1.
Review status: 12 of 15 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/memo.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-eflO2iaZdr2EA_0O:-L3-eflO2iaZdr2EA_0P:b9cakrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo.go#L111)):*
> ```Go
> //   1: [scan a]
> //
> // See the comments in explorer.go for more details.
> ```

You haven't added explorer.go yet

---

*[pkg/sql/opt/xform/memo_group.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-kdnouWTR-SBewKZP:-L3-kdnouWTR-SBewKZQ:bjiau1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_group.go#L21)):*
> ```Go
> )
> 
> type bitmap = util.FastIntSet
> ```

are you using this? If not, remove it.

---

*[pkg/sql/opt/xform/memo_group.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-l30r6Lkl8aT8mrqr:-L3-l30r6Lkl8aT8mrqs:b1pwmbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_group.go#L39)):*
> ```Go
> // on memoExpr for the definition of logical equivalency. In addition, for each
> // required set of physical properties, the group stores the expression that
> // provides those properties at the lowest known cost.
> ```

I don't see where you store the lowest cost expression... (I assume this is just because you haven't added it yet - but make that clear in the comment)

---

*[pkg/sql/opt/xform/metadata.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-m_7zf_yLqyeqCYqH:-L3-mQyj7CScbhqUil-J:b-kyrj7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata.go#L82)):*
> ```Go
> 
> func newMetadata(catalog optbase.Catalog) *Metadata {
> 	return &Metadata{catalog: catalog, labels: make([]string, 1)}
> ```

Add a comment saying that the ColumnIndex = 0 (and therefore the first label) is reserved

---

*[pkg/sql/opt/xform/metadata_test.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-oNnO11qny-O55ajx:-L3-oNnO11qny-O55ajy:bm2wgyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L34)):*
> ```Go
> }
> 
> func (c *testColumn) IsNullable() bool {
> ```

Add a comment above this method (and the ones below) saying something like:
`// IsNullable is part of the optbase.Column interface`

Not essential since this is a test file, but a good habit to get into...

---

*[pkg/sql/opt/xform/metadata_test.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-p4-ShGPykURzl-K-:-L3-p4-ShGPykURzl-K0:bfrqicx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L53)):*
> ```Go
> }
> 
> func (t *testTable) TabName() string {
> ```

Ditto above for the `optbase.Table` interface

---

*[pkg/sql/opt/xform/metadata_test.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-pIUUJyV6aCL4lUrG:-L3-pIUUJyV6aCL4lUrH:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L71)):*
> ```Go
> }
> 
> func (c *testCatalog) FindTable(ctx context.Context, name *tree.TableName) (optbase.Table, error) {
> ```

Ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L3-sft6VbaoJrcYQz4m:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 17 files reviewed at latest revision, 20 unresolved discussions.

---

*[pkg/sql/opt/xform/expr.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-El3lbVepms7XyB3J:-L3-hAe0ZCyDNtiI_ynf:bwjf69j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L32)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It seems like there could be scenarios where there is a tie (e.g., two join orderings for tables that are exactly the same size).  How are ties broken?
</blockquote></details>

It depends on how the optimizer decides to break the tie. The v4 optimizer just keeps whatever Expr it happened to come across first, which is arbitrary, but fine for our purposes.

---

*[pkg/sql/opt/xform/expr.go, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-EISNnromcnlk-4Dk:-L3-iCsDkFwGg66rEZw4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L86)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] The name is not very descriptive, maybe `extractLowestCostExpr`
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/expr.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HTU_2yQhvgOscK09:-L3-iFc02nYzLDuDNss6:b-kvn6qh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L114)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would help to add a comment to explain that this functionality isn't implemented yet.
</blockquote></details>

I got rid of this method, will add it back later when needed.

---

*[pkg/sql/opt/xform/factory.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-N-4HNTAVVnT9-MOE:-L3-jQtn1MKdL2pS1KnR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/factory.go#L32)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Add a comment to explain what `maxSteps` is.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/logical_props.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-GSb8haj05RR1hluL:-L3-kVXzIPYA1kNIY_o_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props.go#L65)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] in opt we switched to use `,null` for nullable columns instead (`*` is cryptic)
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-OxjyMkuiFSMXTtl4:-L3-kccVbCN6j1eygGtI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_factory.go#L38)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] fix spacing issues in the NOTE
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 187 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-S6kBBgInFZNbdMyl:-L3-lGUX0zbYEdW94sNV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_factory.go#L187)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Never mind - I was confused about what the colMap was doing. Might be good to add a comment with a reminder/explanation about what colMap is.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/logical_props_test.go, line 79 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-Ziv4665Mng1UZlYq:-L3-lbFcQObiEPeMn2yp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_test.go#L79)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

when you pass in a constant like 0 or false, it's good practice to add a comment like this:
```
factory := newFactory(gen.mem, 0 /* maxSteps */)
```
It's in the style guide somewhere...
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/logical_props_test.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-_Ahh9-e9uH-5sjB5:-L3-lbgTDt_RJqkl1HeB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/logical_props_test.go#L83)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

ditto here - add a comment for the 0
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-eflO2iaZdr2EA_0O:-L3-tDgizlU-gpPeDKSq:bg72l3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo.go#L111)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You haven't added explorer.go yet
</blockquote></details>

Right, but I figured I'd leave that comment as-is so I don't forget to add it back in.

---

*[pkg/sql/opt/xform/memo_expr.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-Gp4D86is8Idjge5_:-L3-lmoR8TSOgpWnX5i0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] expr.og.go, contains
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo_expr.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HDY3ZXPm5a4iOqaP:-L3-m5CLd2LpjMwHRzwG:be0shf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L52)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

24 bytes (or just "small")
</blockquote></details>

It should be only 16 bytes: Operator is uint16, and the state field is three uint32 values, which means the whole struct should be just 16 bytes. I used unsafe.Sizeof to verify its size on a 64-bit machine.

---

*[pkg/sql/opt/xform/memo_expr.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HVO5xU-y8htbglQ3:-L3-mlW2BrCRP7F2Xu9Z:b-362fel) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L61)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] memoGroupString (maybe it should be a method on memoGroup?)
</blockquote></details>

Agree, this should be on memoGroup, can't remember how it ended up here.

---

*[pkg/sql/opt/xform/memo_group.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-kdnouWTR-SBewKZP:-L3-tQW2uKTsBp1w7kXb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_group.go#L21)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

are you using this? If not, remove it.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo_group.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-l30r6Lkl8aT8mrqr:-L3-tbalj5B35oyv10Zq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_group.go#L39)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I don't see where you store the lowest cost expression... (I assume this is just because you haven't added it yet - but make that clear in the comment)
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/metadata.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-m_7zf_yLqyeqCYqH:-L3-tvfRujrVeh8J2y-b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata.go#L82)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Add a comment saying that the ColumnIndex = 0 (and therefore the first label) is reserved
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/metadata_test.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-oNnO11qny-O55ajx:-L3-uG5zqadMnctLG58v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L34)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Add a comment above this method (and the ones below) saying something like:
`// IsNullable is part of the optbase.Column interface`

Not essential since this is a test file, but a good habit to get into...
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/metadata_test.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-p4-ShGPykURzl-K-:-L3-uGivEU_kI6Rea-1G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L53)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ditto above for the `optbase.Table` interface
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/metadata_test.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-pIUUJyV6aCL4lUrG:-L3-uI04LSnuWhuFFT0W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L71)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/physical_props.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-IzGNSp5Ktvo63mxa:-L3-sIj18E-6uPpWaXEs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/physical_props.go#L42)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a // TODO here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
OK, I've addressed all the comments, and ready for round 2.

---

Review status: 0 of 17 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L3-wTT7I1TX8nNO7dNE:b-2yfgr8)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 14 of 14 files at r2.
Review status: 14 of 17 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/expr.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-El3lbVepms7XyB3J:-L3-wer7WMnEZcMI_94W:b4l94wr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L32)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It depends on how the optimizer decides to break the tie. The v4 optimizer just keeps whatever Expr it happened to come across first, which is arbitrary, but fine for our purposes.
</blockquote></details>

Seems reasonable - maybe add that to the comment?

---

*[pkg/sql/opt/xform/logical_props.go, line 67 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-xGLtl2sETmOhLZ39:-L3-xGLtl2sETmOhLZ3A:b-cj0fpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8fbd9cc55b3fca42b026dd67a46e5040193d86/pkg/sql/opt/xform/logical_props.go#L67)):*
> ```Go
> 			buf.WriteByte(':')
> 			if p.Relational.NotNullCols.Contains(int(colIndex)) {
> 				buf.WriteString("null:")
> ```

If you're doing it this way, change the if statement to:
`if !p.Relational.NotNullCols.Contains(int(colIndex)) {`

(otherwise it should say "not null:")

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 187 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30BgQJGkH0pG0dUwu8:-L30BgQJGkH0pG0dUwu9:b-tdhhlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8fbd9cc55b3fca42b026dd67a46e5040193d86/pkg/sql/opt/xform/logical_props_factory.go#L187)):*
> ```Go
> 
> 	// Columns have to be not-null on both sides to be not-null in result.
> 	// Colmap matches columns on the left side of the operator with columns on
> ```

[nit] `Colmap` -> `colMap`, `colset` -> `ColSet`

---

*[pkg/sql/opt/xform/memo.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-eflO2iaZdr2EA_0O:-L3-y2UgvFKX9_hwZr2S:b-ur7973) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo.go#L111)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Right, but I figured I'd leave that comment as-is so I don't forget to add it back in.
</blockquote></details>

Ok - in that case, maybe mark it with a TODO

---

*[pkg/sql/opt/xform/metadata_test.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-oNnO11qny-O55ajx:-L3-yV0Kd5XmQPSIB6M0:bt20ato) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L34)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks good.  Another common pattern in the cockroach codebase (in addition to the comments) is to add the following line:
`var _ optbase.Column = &testColumn{}`

It helps documentation and provides an extra correctness test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 9 of 17 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/opt/xform/expr.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-El3lbVepms7XyB3J:-L30JiDsi4k2C-_jw79F:b-ju8s0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L32)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Seems reasonable - maybe add that to the comment?
</blockquote></details>

I think that should be something I add to the optimizer comment. I think it's too much detail to put on Expr.

---

*[pkg/sql/opt/xform/logical_props.go, line 67 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-xGLtl2sETmOhLZ39:-L30Jw0LW_qFKT4gnaE_:b-wn61te) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8fbd9cc55b3fca42b026dd67a46e5040193d86/pkg/sql/opt/xform/logical_props.go#L67)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

If you're doing it this way, change the if statement to:
`if !p.Relational.NotNullCols.Contains(int(colIndex)) {`

(otherwise it should say "not null:")
</blockquote></details>

Oh, whoops, nice catch. I even changed the tests to verify the wrong thing! :disappointed: 

---

*[pkg/sql/opt/xform/logical_props_factory.go, line 187 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30BgQJGkH0pG0dUwu8:-L30KdDOxzrxr5GuTgWW:b-u3m43d) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8fbd9cc55b3fca42b026dd67a46e5040193d86/pkg/sql/opt/xform/logical_props_factory.go#L187)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] `Colmap` -> `colMap`, `colset` -> `ColSet`
</blockquote></details>

I changed this to read:

// colMap matches columns on the left side of the operator with columns on
// the right side, since OutputCols are not ordered and may not correspond
// to each other.

---

*[pkg/sql/opt/xform/memo.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-eflO2iaZdr2EA_0O:-L30KsgzEw-BRQunY0Bs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo.go#L111)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ok - in that case, maybe mark it with a TODO
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/metadata_test.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-oNnO11qny-O55ajx:-L30L93C0V8yRbv3zJwU:b74ph9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/metadata_test.go#L34)):*
> var _ optbase.Column = &testColumn{}

Interesting, good to know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Some random comments below, all of which can be addressed in follow-on PRs.

---

Review status: 9 of 17 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.go, line 43 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30j4JLYUWSvVBx6azX:-L30j4JLYUWSvVBx6azY:b-ygy7wp) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/expr.go#L43)):*
> ```Go
> //
> //    +-----------+  +-----------+
> // 3: | join 1, 2 |  | join 2, 1 |
> ```

Random nit (that you can fix in another PR): the debug output from memo and the comments in `memo.go` display the memo contents without ascii box art. 

---

*[pkg/sql/opt/xform/memo.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30n7NDgcy-r8_bo6Zl:-L30n7NDgcy-r8_bo6Zm:baehy87) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L176)):*
> ```Go
> 	mgrp := m.newGroup(norm)
> 	e := extractLowestCostExpr(m, mgrp.id, minPhysPropsID)
> 	mgrp.logical = m.logPropsFactory.constructProps(&e)
> ```

Does the `&e` force `e` to the heap?

---

*[pkg/sql/opt/xform/memo.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30mZyiAT1gRglQZmIH:-L30mZyjfPBLUD3tOeZi:bkdj3ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L200)):*
> ```Go
> // storeList allocates storage for a list of group IDs in the memo and returns
> // an ID that can be used for later lookup.
> // TODO(andyk): lists should be interned.
> ```

Not sure what you had in mind here, but we could do some unsafe trickery and cast `items` to a `[]byte` slice an have a `listMap map[string]ListID`. The lookup in a `map[string]` has an optimization for `map([]byte()]` which avoids an allocation. 

---

*[pkg/sql/opt/xform/memo.go, line 217 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30lu-AlmxzIBtqAyVc:-L30lu-AlmxzIBtqAyVd:btosd6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L217)):*
> ```Go
> // this method is a no-op and returns the ID of the previous value.
> func (m *memo) internPrivate(private interface{}) PrivateID {
> 	id, ok := m.privatesMap[private]
> ```

Note that `privatesMap` restricts the `private` data to types that can be map keys. That means no slices or structs which contain slices. I don't see why this will be burdensome now, but mentioning it so that others are aware if we stumble across this restriction in the future.

---

*[pkg/sql/opt/xform/memo.go, line 234 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30lFH_3JKHB_VB8k3o:-L30lFH_3JKHB_VB8k3p:b-rexhlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L234)):*
> ```Go
> func (m *memo) String() string {
> 	var buf bytes.Buffer
> 	for i := len(m.groups) - 1; i > 0; i-- {
> ```

No topological sort of the groups? Failing that, we should highlight which group is the root, though it seems you're not maintaining that info in the `memo` right now.

---

*[pkg/sql/opt/xform/memo_expr.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HDY3ZXPm5a4iOqaP:-L30jUeXe72XebgTLopl:bvz2vn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L52)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It should be only 16 bytes: Operator is uint16, and the state field is three uint32 values, which means the whole struct should be just 16 bytes. I used unsafe.Sizeof to verify its size on a 64-bit machine.
</blockquote></details>

FYI, https://github.com/rsc/sizeof is very useful. It can show both structure sizes as well as field offsets. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L30niRi8wI1E2DaSHvn:bxexy8l)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 5 of 5 files at r3.
Review status: 14 of 17 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-El3lbVepms7XyB3J:-L30tAZXDUok52lmBSjb:b-hps91i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/expr.go#L32)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think that should be something I add to the optimizer comment. I think it's too much detail to put on Expr.
</blockquote></details>

Ok sounds good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 14 of 17 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.go, line 43 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30j4JLYUWSvVBx6azX:-L31fPqHxlEx9hNZNVdQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/expr.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Random nit (that you can fix in another PR): the debug output from memo and the comments in `memo.go` display the memo contents without ascii box art. 
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30n7NDgcy-r8_bo6Zl:-L31gOEwkGWgvi2aXPSe:bub3lv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L176)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does the `&e` force `e` to the heap?
</blockquote></details>

From the assembly, it didn't look like it to me, though it's a bit hard to tell, since there's a lot of inlining going on, including some allocations done by newGroup. That said, I actually don't care if it gets forced on the heap in this case, since there's a lot more happening here including other allocations. I'm more concerned about other places where we're traversing an entire sub-tree of expressions - I don't want allocations to occur in that case.

---

*[pkg/sql/opt/xform/memo.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30mZyiAT1gRglQZmIH:-L31gzsAuXPVXX8E9Gsn:b-alips7) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L200)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Not sure what you had in mind here, but we could do some unsafe trickery and cast `items` to a `[]byte` slice an have a `listMap map[string]ListID`. The lookup in a `map[string]` has an optimization for `map([]byte()]` which avoids an allocation. 
</blockquote></details>

Yeah, that's a good suggestion. 

---

*[pkg/sql/opt/xform/memo.go, line 217 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30lu-AlmxzIBtqAyVc:-L31gCe-2loUl8dGEoOr:b-vc810a) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L217)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Note that `privatesMap` restricts the `private` data to types that can be map keys. That means no slices or structs which contain slices. I don't see why this will be burdensome now, but mentioning it so that others are aware if we stumble across this restriction in the future.
</blockquote></details>

Yes, that's true. I added a comment about that.

---

*[pkg/sql/opt/xform/memo.go, line 234 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30lFH_3JKHB_VB8k3o:-L31fSp0_eoevqnGxE8a:b-lv1oht) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L234)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

No topological sort of the groups? Failing that, we should highlight which group is the root, though it seems you're not maintaining that info in the `memo` right now.
</blockquote></details>

But the groups are already in a pretty good order, since child groups always must have been added to the memo before their parent. What's the additional benefit of doing an explicit topological sort? Also, as you point out, the root is maintained outside the memo. I didn't see value in keeping it in the memo, though that may change if we find it useful later on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 12 of 17 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/memo.go, line 234 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30lFH_3JKHB_VB8k3o:-L32ha0ynXGMUYdneC9f:bm84u9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L234)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But the groups are already in a pretty good order, since child groups always must have been added to the memo before their parent. What's the additional benefit of doing an explicit topological sort? Also, as you point out, the root is maintained outside the memo. I didn't see value in keeping it in the memo, though that may change if we find it useful later on.
</blockquote></details>

The child groups are only added before the parent when initially adding an expression to the memo (or constructing the memo as in v4). During search, when a transformation creates a new expression that expression will get a higher group number and then display of the memo won't make it obvious which expression is the root and which are children. Topologically sorting addresses both problems and it isn't difficult to do (the code is in `opttoy/v3/memo.go`). For a simple example of a query where a child group has a larger group number than its parent group, see https://github.com/petermattis/opttoy/blob/master/v3/testdata/search#L37-L46:

```
SELECT * FROM a, b, c
----
5: [inner-join [3 4 -]] [inner-join [6 2 -]] [inner-join [7 1 -]] [inner-join [4 3 -]] [inner-join [2 6 -]] [inner-join [1 7 -]]
7: [inner-join [2 4 -]] [inner-join [4 2 -]]
6: [inner-join [1 4 -]] [inner-join [4 1 -]]
4: [scan c] [index-scan c@primary [-]]
3: [inner-join [1 2 -]] [inner-join [2 1 -]]
2: [scan b] [index-scan b@primary [-]]
1: [scan a] [index-scan a@primary [-]]
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 12 of 17 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/memo.go, line 234 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30lFH_3JKHB_VB8k3o:-L335SKMI4W-WjgMG1nR:b-33zoh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L234)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The child groups are only added before the parent when initially adding an expression to the memo (or constructing the memo as in v4). During search, when a transformation creates a new expression that expression will get a higher group number and then display of the memo won't make it obvious which expression is the root and which are children. Topologically sorting addresses both problems and it isn't difficult to do (the code is in `opttoy/v3/memo.go`). For a simple example of a query where a child group has a larger group number than its parent group, see https://github.com/petermattis/opttoy/blob/master/v3/testdata/search#L37-L46:

```
SELECT * FROM a, b, c
----
5: [inner-join [3 4 -]] [inner-join [6 2 -]] [inner-join [7 1 -]] [inner-join [4 3 -]] [inner-join [2 6 -]] [inner-join [1 7 -]]
7: [inner-join [2 4 -]] [inner-join [4 2 -]]
6: [inner-join [1 4 -]] [inner-join [4 1 -]]
4: [scan c] [index-scan c@primary [-]]
3: [inner-join [1 2 -]] [inner-join [2 1 -]]
2: [scan b] [index-scan b@primary [-]]
1: [scan a] [index-scan a@primary [-]]
```
</blockquote></details>

OK, makes sense. I was only considering normalization patterns, which always add nodes in topological order. But once we have exploration patterns, that changes. At that point, we can add in the topological sort code (and the root, if it makes the code nicer/simpler).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 12 of 17 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.og.go, line 0 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L33dfgbydi9VVekq15I:-L33dfgbydi9VVekq15J:byrk6br) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4883c126e9dd56c44d30d4c9c2bd18564183dc/pkg/sql/opt/xform/expr.og.go#L0)):*
[nit] In the future we should make these not show up in diffs. For reviewable https://github.com/Reviewable/Reviewable/wiki/FAQ#how-do-i-tell-reviewable-that-a-file-is-generated-and-should-not-be-reviewed

---

*[pkg/sql/opt/xform/memo_expr.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L3-HDY3ZXPm5a4iOqaP:-L33cgvajKEKNfxHcHkl:brtyl28) ([raw file](https://github.com/cockroachdb/cockroach/blob/b09a9c3d1c2ef7dc897c1ec81986b953c89dbb6b/pkg/sql/opt/xform/memo_expr.go#L52)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

FYI, https://github.com/rsc/sizeof is very useful. It can show both structure sizes as well as field offsets. 
</blockquote></details>

Ah, I was thinking of expr not memoExpr

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L33ebymEQWVlnIVDfVY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/opt/xform/memo.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30n7NDgcy-r8_bo6Zl:-L33gU9YSNZZkrQMzw48:b44ajv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L176)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

From the assembly, it didn't look like it to me, though it's a bit hard to tell, since there's a lot of inlining going on, including some allocations done by newGroup. That said, I actually don't care if it gets forced on the heap in this case, since there's a lot more happening here including other allocations. I'm more concerned about other places where we're traversing an entire sub-tree of expressions - I don't want allocations to occur in that case.
</blockquote></details>

An alternative to looking over the assembly is `go build -gcflags -m` which shows where in the code something is moved to heap. (http://www.agardner.me/golang/garbage/collection/gc/escape/analysis/2015/10/18/go-escape-analysis.html)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 12 of 17 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.og.go, line  at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L33dfgbydi9VVekq15I:-L33fwdo6LbccyA9JQ2g:broyrrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4883c126e9dd56c44d30d4c9c2bd18564183dc/pkg/sql/opt/xform/expr.og.go#L)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] In the future we should make these not show up in diffs. For reviewable https://github.com/Reviewable/Reviewable/wiki/FAQ#how-do-i-tell-reviewable-that-a-file-is-generated-and-should-not-be-reviewed
</blockquote></details>

Good idea, though I don't know why reviewable doesn't already understand the standard go format (https://github.com/golang/go/issues/13560). We'd have to add another line to our files, since Go's syntax is not compatible with reviewable's.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 12 of 17 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/memo.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L30n7NDgcy-r8_bo6Zl:-L33iKgAdr6KWGsJFvWv:b43m9b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b7e3587a958b7a25c23e3b7c4c0768739a0caf/pkg/sql/opt/xform/memo.go#L176)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

An alternative to looking over the assembly is `go build -gcflags -m` which shows where in the code something is moved to heap. (http://www.agardner.me/golang/garbage/collection/gc/escape/analysis/2015/10/18/go-escape-analysis.html)
</blockquote></details>

Nice, I hadn't known about that tool. According to the tool, the &e *does* escape to the heap. In this context, I'm not too worried about that, though I suppose we could pass e by value to prevent that if we're concerned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Curious to see the optgen files — any reason they're not checked in?

andy-kimball commented:
Because I need to go through them, get them cleaned up and commented, etc. before checking them in. That's going to take some time, and I wanted to unblock Becca before then.

vilterp commented:
Gotcha. Exciting work!

tschottdorf commented:
Left some comments, but they're all take it or leave it.

---

Review status: 12 of 17 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.go, line 28 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L33sRKcb8cgq2SKdyXl:-L33sRKcb8cgq2SKdyXm:bbq4bj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4883c126e9dd56c44d30d4c9c2bd18564183dc/pkg/sql/opt/xform/expr.go#L28)):*
I looked into `memo.go` but I'm still not sure what `virtualizing` is. I think it's what's described later in the comment but a definition wouldn't hurt. My use below is likely off.

> Expr provides a view of the (unique) tree in the memo forest which has the requested physical properties (called "virtualizing the forest") making it both simple and efficient to traverse that tree and examine its properties.

Bonus points for motivating the word `virtualizing`, which I haven't quite managed for myself (and I wasn't able to grep any other uses of it).

---

*[pkg/sql/opt/xform/logical_props.go, line 29 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L33vMUUcQvBKQmPdWHo:-L33vMUUcQvBKQmPdWHp:blqcg9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4883c126e9dd56c44d30d4c9c2bd18564183dc/pkg/sql/opt/xform/logical_props.go#L29)):*
> ```Go
> // result using different algorithms, the complete set of data is returned
> // and can be transformed into whatever layout or presentation format that is
> // is desired.
> ```

double `is`.

---

*[pkg/sql/opt/xform/memo.go, line 32 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L33wH8p74aD05IUisN5:-L33wH8qwzZ7Yo0v9UxM:b-24ms2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4883c126e9dd56c44d30d4c9c2bd18564183dc/pkg/sql/opt/xform/memo.go#L32)):*
> ```Go
> // by the memo. The ID consists of an offset into the memo's lists slice, plus
> // the number of elements in the list. Lists have numbers greater than 0; a
> // ListID of 0 indicates an undefined list (probable indicator of a bug).
> ```

// Lists have offsets and lengths greater than 0; a ListID of `{0, 0}` indicates an undefined list (probable indicator of a bug).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-:-L35tq6LYPRJV9RvvAR9:b-6i07vf)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 12 of 17 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/expr.go, line 28 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478#-L33sRKcb8cgq2SKdyXl:-L36OaBIvuSqotfcYl4J:bv9x5no) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4883c126e9dd56c44d30d4c9c2bd18564183dc/pkg/sql/opt/xform/expr.go#L28)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I looked into `memo.go` but I'm still not sure what `virtualizing` is. I think it's what's described later in the comment but a definition wouldn't hurt. My use below is likely off.

> Expr provides a view of the (unique) tree in the memo forest which has the requested physical properties (called "virtualizing the forest") making it both simple and efficient to traverse that tree and examine its properties.

Bonus points for motivating the word `virtualizing`, which I haven't quite managed for myself (and I wasn't able to grep any other uses of it).
</blockquote></details>

This gives me an idea. Maybe I should rename `Expr` to be `ExprView`. Is that a more intuitive name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21478)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Adding View to the name definitely helps get my mind on the right track!
-- 

-- Tobias
